File: khf_3_prolog/khf3.pl
% :- pred megoldase(sspec::in, ssol::in).
% megoldase(+SSpec,+SSol) : sikeres, ha az SSol érték-mátrix megoldása az SSpec Sudoku-feladványnak.
megoldase(s(PARAM, MATRIX), SOLUTION) :-
    check(s(PARAM, MATRIX), SOLUTION), !.
megoldase(_,_) :-
    fail,!.

check(s(PARAM, MATRIX), SOLUTION) :-
    normalizeInput(MATRIX, NORMALIZED).

% Returns Rth row
getRow(MATRIX, R, ROW) :-
    getNth(MATRIX, R, ROW).

% Returns Cth column
getCol(MATRIX, C, COL) :-
    getCol(MATRIX, C, [], COL).
getCol([ROW|TAIL], C, TEMP, COL) :-
    getNth(ROW, C, ELEMENT),
    append(TEMP, [ELEMENT], TEMP1),
    getCol(TAIL, C, TEMP1, COL).
getCol([], _, COL, COL).

% Returns the NTH element from list
getNth([HEAD|_], 1, HEAD).
getNth([_|TAIL], NTH, RET) :-
    NEXT is NTH-1,
    getNth(TAIL, NEXT, RET).

% Examine list
examine(NORMALIZED, NTH, RESTRICTED) :-
    examine(NORMALIZED, NTH, [], RESTRICTED).
examine([[NUMBER,_,_,_]|TAIL], NTH, TEMP, RESTRICTED) :-
    NTH=:=1 ->
        NEXT is NTH-1,
        examine(TAIL, NEXT, TEMP, RESTRICTED);
    %else
        NUMBER> -1 ->
            NEXT is NTH-1,
            append(TEMP, [NUMBER], TEMP1),
            examine(TAIL, NEXT, TEMP1, RESTRICTED);
    %else
        NEXT is NTH-1,
        examine(TAIL, NEXT, TEMP, RESTRICTED).
examine([], _, RESTRICTED, RESTRICTED).

% Return an item from given matrix structure
getItem([HEAD|TAIL], ROW, COL, RET) :-
    ROW=:=1 ->
        getItem(HEAD, COL-1, RET);
    ROW>0 ->
        getItem(TAIL, ROW-1, COL, RET).
getItem([HEAD|TAIL], COL, RET) :-
    COL=:=0 ->
        RET = HEAD;
    COL>0 ->
        getItem(TAIL, COL-1, RET).

% Create Indexes
getIndex(ROW, COL, PARAM, RET) :-
    TEMP=[],
    getIndex(ROW, COL, PARAM, 0, TEMP, RET).
getIndex(ROW, COL, PARAM, INDEX, TEMP, RET) :-
    PARAM2 is PARAM*PARAM,
    INDEX=:=PARAM2 ->
        RET = TEMP,
        true;
    mod(COL, PARAM, X),
    X=:=0 ->
        ROW1 is ROW+1,
        COLPARAM is COL-PARAM+1,
        INDEX1 is INDEX+1,
        append(TEMP, [[ROW, COL]], TEMP1),
        getIndex(ROW1, COLPARAM, PARAM, INDEX1, TEMP1, RET);
    %else
        COL1 is COL+1,
        INDEX1 is INDEX+1,
        append(TEMP, [[ROW, COL]], TEMP1),
        getIndex(ROW, COL1, PARAM, INDEX1, TEMP1, RET).

% Swap indexes to elements from MATRIX
swapIndexesForElements(INDEXMATRIX, MATRIX, RET) :-
    TEMP=[],
    swapIndexesForElements(INDEXMATRIX, MATRIX, TEMP, RET).
swapIndexesForElements([[ROW,COL]|TAIL], MATRIX, TEMP, RET) :-
    getItem(MATRIX, ROW, COL, ELEMENT),
    append(TEMP, [ELEMENT], TEMP1),
    swapIndexesForElements(TAIL, MATRIX, TEMP1, RET).
swapIndexesForElements([], _, TEMP, RET) :-
    RET = TEMP,
    true.

% Create PARAM*PARAM size squares from MATRIX
makeArrays(MATRIX, PARAM, RET) :-
    TEMP=[],
    makeArrays(MATRIX, PARAM, TEMP, 1, 1, RET).
makeArrays(MATRIX, PARAM, TEMP, ROWK, COLK, RET):-
    ROWK>PARAM ->
        RET=TEMP,
        true;
    COLK=:=PARAM ->
        ROWK1 is ROWK+1,
        makeArraysHelper(ROWK, COLK, PARAM, MATRIX, TEMP, TEMP1),
        makeArrays(MATRIX, PARAM, TEMP1, ROWK1, 1, RET);
    %else
        COLK1 is COLK+1,
        makeArraysHelper(ROWK, COLK, PARAM, MATRIX, TEMP, TEMP1),
        makeArrays(MATRIX, PARAM, TEMP1, ROWK, COLK1, RET).
makeArraysHelper(ROWK, COLK, PARAM, MATRIX, IN, RET) :-
    ROWB is 1+(ROWK-1)*PARAM,
    COLB is 1+(COLK-1)*PARAM,
    getIndex(ROWB, COLB, PARAM, INDEXMATRIX),
    swapIndexesForElements(INDEXMATRIX, MATRIX, ELEMENTMATRIX),
    append(IN, [ELEMENTMATRIX], RET).

% Change items to tuples {num, even/odd, south/west}
normalizeInput(MATRIX, RET) :-
    TEMP=[],
    normalizeRows(MATRIX, TEMP, RET).
normalizeRows([ROW|TAIL], TEMP, RET) :-
    normalizeRow(ROW, NORMALIZED),
    append(TEMP, [NORMALIZED], TEMP1),
    normalizeRows(TAIL, TEMP1, RET).
normalizeRows([], TEMP, TEMP).
normalizeRow(ROW, RET) :-
    TEMP=[],
    normalizeRow(ROW, ROW, TEMP, RET).
normalizeRow([HEAD|TAIL], ROW, TEMP, RET) :-
    createTuple(HEAD, ELEMENT),
    append(TEMP, [ELEMENT], TEMP1),
    normalizeRow(TAIL, ROW, TEMP1, RET).
normalizeRow([], _, TEMP, RET) :-
    RET = TEMP,
    true.

% Create tuple from an 'item' (field)
createTuple(ITEM, [NUMBER, PARITY, SOUTH, WEST]) :-
    getEvenOrOdd(ITEM, PARITY),
    isSouth(ITEM, SOUTH),
    isWest(ITEM, WEST),
    getNumber(ITEM, NUMBER), true.

% True if value satisfy the input form v(value)
isValue(v(_)).

% Gives back the number if exists in item, else -1
getNumber([HEAD|TAIL], NUMBER) :-
    isValue(HEAD) ->
        getNumberHelper(HEAD, NUMBER),
        true;
    %else
    getNumber(TAIL, NUMBER).
getNumber([HEAD], NUMBER) :-
    isValue(HEAD) ->
        getNumberHelper(HEAD, NUMBER);
    %else
        NUMBER = -1.
getNumber([], -1).
getNumberHelper(v(NUM), NUM).

% Gives back proper direction, if given, else return u (unknown)
isSouth(ITEM, SOUTH) :-
    member(s, ITEM) ->
        SOUTH=true;
    SOUTH=false.
isWest(ITEM, WEST) :-
    member(w, ITEM) ->
        WEST=true;
    WEST=false.

% Gives back proper parity, if given, else return u (unknown)
getEvenOrOdd(ITEM, PARITY) :-
    member(o, ITEM) ->
        PARITY=o;
    member(e, ITEM) ->
        PARITY=e;
    %else
        PARITY=u.

% Modulo function
mod(X, Y, RET) :-
    X=:=0 ->
        RET is 0;
    X>0 ->
        RET is X rem Y;
    %else
        RET is Y + X rem Y.

File: khf_3_erlang/khf3.erl
-module(khf3).
-author('falusi.david94@gmail.com').
-vsn('year-mm-dd').
-export([megoldase/2]).

%% @spec khf3:megoldase(SSpec::sspec(), SSol::ssol()) -> B::bool().
%% B igaz, ha SSol megoldĂĄsa az SSpec feladvĂĄnynak.
megoldase({PARAM, MATRIX}, SOLUTION) ->
    NORMALIZED = normalizeInputList({PARAM, MATRIX}),
    NORMALIZEDSOLUTIN = normalizeInputList({PARAM, SOLUTION}),
    %io:format(" ~p", [SOLUTION]),
    %io:format(" ~p", [getCols(SOLUTION, PARAM)]),
    checkRows(SOLUTION, PARAM) andalso
    checkCols(SOLUTION, PARAM) andalso
    checkIntegrity(NORMALIZED, SOLUTION).

checkIntegrity([ROW1| TAIL1], [ROW2|TAIL2]) ->
    RET = checkRowIntegrity(ROW1, ROW2),
    if
        RET ->
            checkIntegrity(TAIL1, TAIL2);
        true ->
            false
    end;
checkIntegrity([], []) -> true .

checkRowIntegrity([[NUMBER, PARITY, WEST, SOUTH]|TAIL1], [VALUE|TAIL2]) ->
    RET =
        (NUMBER=:= -1 orelse NUMBER=:=VALUE) andalso
        (PARITY=:=u
            orelse (PARITY=:=e andalso mod(VALUE, 2)=:=0)
            orelse (PARITY=:=o andalso mod(VALUE, 2)=:=1)),
    if
        RET ->
            checkRowIntegrity(TAIL1, TAIL2);
        true ->
            false
    end;
checkRowIntegrity([], []) -> true.

checkRows([SOLROW|TAIL], PARAM) ->
    RET = checkFullList(SOLROW, PARAM),
    if
        RET ->
            checkRows(TAIL, PARAM);
        true ->
            false
    end;
checkRows([], _) -> true.

checkCols(MATRIX, PARAM) ->
    COLS = getCols(MATRIX, PARAM),
%    io:format(" ~p", [COLS]),
    checkRows(COLS, PARAM).

% Get columns from SOLUTION matrix
getCols(MATRIX, PARAM) ->
    getCols(MATRIX, 1, PARAM*PARAM, []).
getCols(MATRIX, COL, PARAM2, TEMP) ->
    if
        PARAM2<COL ->
            TEMP;
        true ->
            COLUMN = getCol(MATRIX, 1, COL, PARAM2, []),
            getCols(MATRIX, COL+1, PARAM2, lists:append(TEMP, [COLUMN]))
    end.
getCol(MATRIX, ROW, COL, PARAM2, TEMP) ->
    if
        PARAM2<ROW ->
            TEMP;
        true ->
            ITEM = getItem(MATRIX, ROW, COL),
            getCol(MATRIX, ROW+1, COL, PARAM2, lists:append(TEMP, [ITEM]))
    end.

% Check if every number from 1..PARAM*PARAM is in the list
checkFullList(LIST, PARAM) ->
    checkFullListHelper(LIST, PARAM*PARAM).
checkFullListHelper(LIST, INDEX) ->
    RET = lists:member(INDEX, LIST),
    if
        RET ->
            checkFullListHelper(LIST, INDEX-1);
        INDEX<1 ->
            true;
        true ->
            false
    end.

% Select the needed square from squares
getSquare([HEAD|_], 1) ->
    HEAD;
getSquare([_|TAIL], INDEX) ->
    getSquare(TAIL, INDEX-1).

% Change items to tuples {num, even/odd, south/west}
normalizeInputList({K, LIST}) ->
    normalizeInputList({K, LIST}, 1, 1, [], []).
normalizeInputList({K, LIST}, ROWI, COLI, RET, TEMP) ->
    if
        ROWI>K*K ->
            RET;
        COLI>K*K ->
            normalizeInputList({K, LIST}, ROWI+1, 1, lists:append(RET, [TEMP]), []);
        true ->
            ITEM = createTuple(getItem(LIST, ROWI, COLI)),
            normalizeInputList({K, LIST}, ROWI, COLI+1, RET, lists:append(TEMP, [ITEM]))
    end.

% Create tuple from an 'item' (field)
createTuple(ITEM) ->
    PARITY = evenOrOdd(ITEM),
    NUMBER = number(ITEM),
    [NUMBER, PARITY, lists:member(w, ITEM), lists:member(s, ITEM)].

% Gives back the number if exists in item, else -1
number([HEAD]) ->
    if
        is_number(HEAD) ->
            HEAD;
        true ->
            -1
    end;
number([HEAD|TAIL]) ->
    if
        is_number(HEAD) ->
            HEAD;
        true ->
            number(TAIL)
    end;
number([]) ->
    -1.

% Gives back proper parity, if given, else return u (unknown)
evenOrOdd(ITEM) ->
    MEMBER_O = lists:member(o, ITEM),
    MEMBER_E = lists:member(e, ITEM),
    if
        MEMBER_O ->
            o;
        MEMBER_E ->
            e;
        true ->
            u
    end.

% Return an item from given matrix structure
getItem([HEAD|TAIL], ROWI, COLI, ROW, COL) ->
    ELEMENT_LIST = isElementList(HEAD),
    if
        is_list(HEAD) andalso ELEMENT_LIST ->
            if
                ROWI<ROW ->
                    getItem(TAIL, ROWI+1, COLI, ROW, COL);
                ROWI=:=ROW ->
                    getItem(HEAD, ROWI, COLI, ROW, COL)
            end;
        is_list(HEAD) andalso not ELEMENT_LIST ->
            if
                COLI<COL ->
                    getItem(TAIL, ROWI, COLI+1, ROW, COL);
                COLI=:=COL ->
                    getItem(HEAD, ROWI, COLI, ROW, COL)
            end;
        true ->
            [HEAD|TAIL]
    end;
getItem([], _, _, _, _) ->
    [].
getItem([HEAD|TAIL], ROW, COL) ->
    getItem([HEAD|TAIL], 1, 1, ROW, COL).

% Returns true if the first element is a list of the input list,
% every other case returns false
isElementList([HEAD|_]) ->
    is_list(HEAD);
isElementList(_) ->
    false.

% Create
cutIntoArrays(LIST, K) ->
    cutIntoArrays(LIST, K, K, 1, 1, 1, 1, K*K, K*K, 1, [], []).
cutIntoArrays(L, R, C, RI, CI, J, K, CMAX, RMAX, ASD, LIST, RET) ->
    NEW = (mod(ASD,R*C) =:= 0) or
            ( ((CI =:= CMAX) and (mod(RI,R) =:= 0)) or ((RI =:= RMAX) and (mod(CI,C) =:= 0)))
            or (RI*CI=:=RMAX*CMAX),
    if
        NEW ->
            ELEMENT = getItem( L, RI, CI),
            LIST1 = [],
            ASD1=0,
            RET1 = lists:append( RET, [lists:append(LIST, [ELEMENT])]);
        true ->
            RET1 = RET,
            ASD1=ASD,
            ELEMENT = getItem( L, RI, CI),
            LIST1 = lists:append(LIST, [ELEMENT])
    end,
    if
        CI=:=CMAX ->
            if
                RI=:=RMAX ->
                    RET1;
                RI=:=(K*R) ->
                    cutIntoArrays(L, R, C, RI+1, 1, 1, K+1, CMAX, RMAX, ASD1+1, LIST1, RET1);
                true ->
                     cutIntoArrays(L, R, C, RI+1, (J-1)*C+1, J, K, CMAX, RMAX, ASD1+1, LIST1, RET1)
             end;
        CI=:=(J*C) ->
            if
                RI=:=RMAX ->
                    cutIntoArrays(L, R, C, (K-1)*R+1, CI+1, J+1, K, CMAX, RMAX, ASD1+1, LIST1, RET1);
                RI=:=(K*R) ->
                    cutIntoArrays(L, R, C, (K-1)*R+1, CI+1, J+1, K, CMAX, RMAX, ASD1+1, LIST1, RET1);
                true ->
                    cutIntoArrays(L, R, C, RI+1, (J-1)*C+1, J, K, CMAX, RMAX, ASD1+1, LIST1, RET1)
            end;
         true ->
             cutIntoArrays(L, R, C, RI, CI+1, J, K, CMAX, RMAX, ASD1+1, LIST1, RET1)
    end.

mod(X,Y) when X > 0 -> X rem Y;
mod(X,Y) when X < 0 -> Y + X rem Y;
mod(0,_) -> 0.
